APIT-130 improve workflow

folkengine opened the pull request:
This PR is to update the `everest-utils` repository dev container environment based upon updates discovered by the dev team in order to be able to use the EVerest Playground inside VSCode. 

It includes:

* Updates to `docker/docker-compose.yml` so that the environment can properly connect to the local docker network.
* Updates to the `docker/everest-playground/Dockerfile` to extend the timeout for the environment.
* Updates to the `everest-cpp/init.sh` script installing the Python dependencies for the `Josev` (needed to run the Node-RED env) and `everest-testing` subprojects. 
* `docs/MAC_SETUP.md` documenting how to setup your mac to use the environment.

NOTE: At first I thought that the reason we were seeing issues was because of differences in using a Mac for development, however, I confirmed that they same issues are there on Linux and that the updates fix the problems there as well.

folkengine commented:
@shankari please review

shankari commented on docs/MAC_SETUP.md:
````diff
@@ -0,0 +1,103 @@
+$> bin/devup
+```
+
+### Create docker network
+
+```
+docker network create --driver bridge --ipv6  --subnet fd00::/80 infranet_network --attachable
+```
+
````
I generally don't like replicating the steps from a script in the README because then there is not a single source of truth, and we have to ensure that the two are in sync.

Having said that, maybe this is just my resource-strapped open source project experience speaking, and better resources projects will devote the time to keeping them in sync.

Although this is also an open-source project... 😄

folkengine replied:
I'm actually in agreement with you on this one. My preference is to have any of these in a scripts directory on path to be referenced from the README. My worry is that this isn't the current EVerest way of doing things, and that maybe it's better that I just remove the scripts altogether, no matter how convenient I find them.

folkengine replied:
I've removed the script so that there's no confusion.

shankari commented on docs/MAC_SETUP.md:
```diff
@@ -0,0 +1,103 @@
+
+In VSCode:
+
+* Press `CMD + Shift + P`
+* type `Dev Containers: Open Folder in Container...`
+* Navigate and open the docker/everest-playground folder. 
+
+This will open the EVerest Playground as a VSCode dev container. You should now be ready to go.
```
I have never used dev containers before - I edit my code using vim 😄 - but this does seem to be very useful for those who use vscode. But doesn't this need a `devcontainer.json` to work? I don't see the file in the PR.

Or do we not need it because we don't need any additional dependencies/plugins?

folkengine replied:
The [`devcontainer.json`](https://github.com/US-JOET/everest-utils/blob/APIT-130_improve_workflow/docker/everest-playground/.devcontainer.json) file is in the same folder as the [`Dockerfile`](https://github.com/US-JOET/everest-utils/tree/APIT-130_improve_workflow/docker/everest-playground) that we're opening up.

folkengine replied:
Respect 😄. When I grow up, I want to be a neovim developer.

shankari commented on docs/MAC_SETUP.md:
````diff
@@ -0,0 +1,103 @@
+
+This will open the EVerest Playground as a VSCode dev container. You should now be ready to go.
+
+## Initialize 
+
+Now we run commands inside the Playground based on the instructions from the [`everest-cpp`](/everest-cpp/README.md)
+subdirectory.
+
+Initialize the EVerest workspace sourcing the *[init.sh](./init.sh)* file:
+
+```bash
+. init.sh
+```
````
These instructions have to be run manually, correct? If we don't refactor to move the previous instructions out of the README, we should clarify that the `devup` script covers everything until here.

folkengine replied:
I removed the `devup` script so that there's no confusion.

shankari commented on docs/MAC_SETUP.md:
````diff
@@ -0,0 +1,103 @@
+cmake .. && make install -j12
+```
+
+### Starting the NodeRed Environment
+
+```bash
+./dist/bin/manager --config ../config/config-sil-dc.yaml
+```
+
````
I think this can be clarified better.

- The `manager` is the everest manager
- Starting it with the `config-sil-dc.yml` sets up the modules that are loaded in EVerest
- The node-red environment is just for the UI to interact with the simulation.

```suggestion
### Starting everest with SIL simulation and a user interface

```bash
./dist/bin/manager --config ../config/config-sil-dc.yaml
```

```

shankari reviewed (changes requested):
@folkengine would it also be possible to add some screenshots/script outputs of the "Testing done" to clarify what we expect to see. I think it really helps to give reviewers a sense of what the outcome of the merge would be!

folkengine commented:
Yes, that's a great idea.

folkengine commented:
Added screenshots to MAC_SETUP.

shankari reviewed (approved):
Aw... I am a big fan of setup scripts, and my thought from the review was that you would keep devup.sh and simplify the README to refer to it. The explanations here would be in the script.

I am fine with merging this as is on the principle that it is better than what we have now, but I would encourage you to create a new PR with the devup script and see what response it gets from the community. I will definitely be supportive!

folkengine commented:
> Aw... I am a big fan of setup scripts, and my thought from the review was that you would keep devup.sh and simplify the README to refer to it. The explanations here would be in the script.
> 
> I am fine with merging this as is on the principle that it is better than what we have now, but I would encourage you to create a new PR with the devup script and see what response it gets from the community. I will definitely be supportive!

That's great to hear. I'll add them back.

folkengine commented:
> Aw... I am a big fan of setup scripts, and my thought from the review was that you would keep devup.sh and simplify the README to refer to it. The explanations here would be in the script.
> 
> I am fine with merging this as is on the principle that it is better than what we have now, but I would encourage you to create a new PR with the devup script and see what response it gets from the community. I will definitely be supportive!

@shankari I've added back the script and updated the README to point to it.

shankari reviewed (approved):
I think this is fine. My only concern is whether people will find these instructions, given that they are tucked away into `everest-utils/docs`. I would suggest moving to a more prominent location after checking with the community,

folkengine commented:
It included a custom environment variable that I thought made it too complex. Didn't add any real value IMNSHO. 

Moving it is a good call.

andistorm commented on bin/devup:
```diff
@@ -0,0 +1,22 @@
+echo "Standing up Dev Everest Docker Playground"
+
+docker compose -f "./everest-utils/docker/docker-compose-dev.yml" up -d mqtt-server 
+docker compose -f "./everest-utils/docker/docker-compose-dev.yml" up -d nodered
```
```suggestion
docker compose -f "./everest-utils/docker/docker-compose-dev.yml" up -d nodered

```

andistorm reviewed (changes requested):
LGTM in general.

Adding documentation is always very helpful, thanks!
Our documentation strategy is at the moment to move as much as possible to the main documentation in the `EVerest/EVerest` repository. We already reference it in the README.md https://github.com/EVerest/everest-utils/blob/d03986078c292a84c53076c3c834f05beb7389fb/README.md?plain=1#L5

Documentation regarding devcontainers probably belongs to a new section in [EVerest development tools](https://github.com/EVerest/everest-utils/blob/d03986078c292a84c53076c3c834f05beb7389fb/README.md?plain=1#L5) or to a new section in [Prepare Your Development Environment](https://everest.github.io/nightly/general/02_detail_pre_setup.html)

Could you please move the nice documentation? @krealyt can help you there, if you are not sure where exactly to put it

folkengine commented:
@krealyt can you help me with where this documentation should go? thanks!

krealyt commented:
I appreciate that docs very much, thanks a lot!

I suggest to put them here in a new directory called "how_to_mac" or something like that:
https://github.com/EVerest/EVerest/tree/main/docs/tutorials

I can link the guide from the prerequisites page then.

folkengine commented:
> I appreciate that docs very much, thanks a lot!
> 
> I suggest to put them here in a new directory called "how_to_mac" or something like that: https://github.com/EVerest/EVerest/tree/main/docs/tutorials
> 
> I can link the guide from the prerequisites page then.

@krealyt done!
https://github.com/EVerest/EVerest/pull/156

andistorm reviewed (approved):
LGTM
